src/mrinufft/io/pulseq.py

```diff
@@ -0,0 +1,127 @@
+"""Pulseq Trajectory Reader and writers.
+
+Reads Pulseq `.seq` files to extract k-space trajectory and other parameters. It also provides functionality to create pulseq block and shape objects to
+facilitate the integration of arbitrary k-space trajectories into Pulseq
+sequences. Requires the `pypulseq` package to be installed.
+"""
+
+from types import SimpleNamespace
+import numpy as np
+import pypulseq as pp
+from numpy.typing import NDArray
+from ..trajectories.utils import convert_trajectory_to_gradients
+
+
+def read_pulseq_traj(filename, traj_delay=0.0, grad_offset=0.0):
+    """Extract k-space trajectory from a Pulseq sequence file.
+
+    The sequence should be a valid Pulseq `.seq` file, with arbitrary gradient
+    waveforms, which all have the same length.
+
+    Note
+
+    Parameters
+    ----------
+    filename : str
+        Path to the Pulseq `.seq` file.
+
+    Returns
+    -------
+    description : dict
+        the [DESCRIPTION] block from the sequence file.
+    np.ndarray
+        The k-space trajectory as a numpy array of shape (n_shots, n_samples, 3),
+        where the last dimension corresponds to the x, y, and z coordinates in k-space.
+
+    """
+
+    seq = pp.Sequence()
+    seq.read(filename)
+
+    kspace_adc, kspace, t_exc, t_refocus, t_adc = seq.calculate_kspace(
+        traj_delay, grad_offset
+    )
+
+    # split t_adc with t_exc and t_refocus, the index are then used to split kspace_adc
+
+    t_splits = sorted(np.concatenate([t_exc, t_refocus]))
+    idx_prev = 0
+    kspace_shots = []
+    for t in t_splits:
+        idx_next = np.searchsorted(t_adc, t, side="left")
+        if idx_next == idx_prev:
+            continue
+        if idx_next == len(kspace_adc) and t > t_adc[-1]: # last useful point
+            break
+        kspace_shots.append(kspace_adc[idx_prev:idx_next, :])
+        idx_prev= idx_next
+    kspace_shots = np.asarray(kspace_shots)
+    return kspace_shots
+
+
+def pulseq_grad_blocks(
+    trajectory: NDArray, system: pp.Opts = pp.Opts.default
+) -> list[SimpleNamespace]:
+    """Create Pulseq gradient blocks from a k-space trajectory.
+
+    Parameters
+    ----------
+    trajectory : np.ndarray
+        The k-space trajectory as a numpy array of shape (n_shots, n_samples, 3),
+        where the last dimension corresponds to the x, y, and z coordinates in k-space.
+
+    Returns
+    -------
+    list
+        A list of Pulseq gradient blocks corresponding to the k-space trajectory.
+    """
+    grads = convert_trajectory_to_gradients(trajectory)
+
+    # TODO add prewind/postwind stuff from #276
+    # TODO ensure amplitudes are correct
+
+    # TODO deduplicate gradients (reduces the size of the file, speed up the sequence loading).
+    # TODO Are rotation of trajectories supported in pulseq?
```

Comment:
yes :) https://github.com/imr-framework/pypulseq/blob/master/src/pypulseq/rotate.py
But i wonder why you need it. The rotation of trajectory to match the GOV rotation is done by the sequence at scanner.